File: backend/src/main/java/com/barkowal/nundine/services/ProductService.java
package com.barkowal.nundine.services;

import com.barkowal.nundine.domain.dtos.product.CreateProductRequest;
import com.barkowal.nundine.domain.dtos.product.UpdateProductRequest;
import com.barkowal.nundine.domain.dtos.product.UpdateProductStockRequest;
import com.barkowal.nundine.domain.entities.Product;
import com.barkowal.nundine.domain.entities.ProductStock;

import java.util.List;
import java.util.UUID;

public interface ProductService {
    public Product createProduct(UUID supplier, CreateProductRequest createProductRequest);
    public List<Product> getProducts(UUID userId);
    public Product updateProduct(UUID supplier, UpdateProductRequest updateProductRequest);
    public void deleteProduct(UUID productId, UUID userId);
    public ProductStock getProductStock(UUID productId, UUID inventoryId);
    public void updateProductStock(UUID productId, UUID inventoryId, UpdateProductStockRequest updateProductStockRequest);
}

File: frontend/src/app/services/product-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { BuyProductRequest, CreateProductRequest, UpdateProductRequest, UpdateProductStockRequest } from '../models/Product';

@Injectable({
  providedIn: 'root',
})
export class ProductService {
  private http = inject(HttpClient);
  private baseUrl = 'http://localhost:8080/api/v1';

  createProduct(req: CreateProductRequest): Observable<any> {
    return this.http.post(`${this.baseUrl}/product`, req);
  }

  getProducts(): Observable<any> {
    return this.http.get(`${this.baseUrl}/product`);
  }

  getUserProducts(userId: string | null): Observable<any> {
    return this.http.get(`${this.baseUrl}/product?userId=${userId}`);
  }

  updateProduct(productId: string, req: UpdateProductRequest): Observable<any> {
    return this.http.put(`${this.baseUrl}/product/${productId}`, req);
  }

  deleteProduct(productId: string) {
    return this.http.delete(`${this.baseUrl}/product/${productId}`);
  }

  getProductsStock(): Observable<any> {
    return this.http.get(`${this.baseUrl}/product/stock`);
  }

  updateProductStock(productId: string, req: UpdateProductStockRequest): Observable<any> {
    return this.http.put(`${this.baseUrl}/product/stock/${productId}`, req);
  }

  getShopProducts(): Observable<any> {
    return this.http.get(`${this.baseUrl}/product/shop`);
  }

  buyProduct(req: BuyProductRequest): Observable<any> {
    return this.http.post(`${this.baseUrl}/order`, req)
  }
}
